clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala

```diff
@@ -1,63 +1,254 @@
+import org.broadinstitute.clio.client.dispatch.CommandDispatch
 import org.broadinstitute.clio.client.util.IoUtil
+import org.broadinstitute.clio.client.webclient.ClioWebClient
 
 import akka.actor.ActorSystem
+import akka.http.scaladsl.model.HttpResponse
 import akka.http.scaladsl.model.headers.OAuth2BearerToken
-import caseapp._
+import caseapp.core.{Messages, WithHelp}
 import com.typesafe.scalalogging.LazyLogging
 
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
-object ClioClient
-    extends CommandAppWithPreCommand[CommonOptions, CommandType]
-    with LazyLogging {
-  override val progName = "clio-client"
-  override val appName = "Clio Client"
+/**
+  * The command-line entry point for the clio-client.
+  *
+  * We want most actual logic to happen inside this
+  * object's companion class for maximum testability,
+  * so this object should only handle:
+  *
+  *   1. Setting up and tearing down the actor system
+  *      infrastructure needed by the client.
+  *
+  *   2. Exiting the program with appropriate return
+  *      codes.
+  */
+object ClioClient extends LazyLogging {
+
+  /**
+    * Model representing termination of the client before
+    * it even gets to running a sub-command.
+    *
+    * This could be because of an input error, or because
+    * the user asked for help / usage.
+    *
+    * We model the two cases differently so we can exit with
+    * different return codes accordingly.
+    */
+  sealed trait EarlyReturn
+  final case class ParsingError(error: String) extends EarlyReturn
+  final case class UsageOrHelpAsked(message: String) extends EarlyReturn
 
-  override val appVersion: String = ClioClientConfig.Version.value
+  val progName = "clio-client"
 
-  //not sure this is the best way to pass common opts but case-app docs are spotty at best
-  var commonOptions: CommonOptions = _
+  def main(args: Array[String]): Unit = {
+    implicit val system: ActorSystem = ActorSystem(progName)
+    import system.dispatcher
+    sys.addShutdownHook({ val _ = system.terminate() })
 
-  override def beforeCommand(options: CommonOptions,
-                             remainingArgs: Seq[String]): Unit = {
-    commonOptions = options
-    checkRemainingArgs(remainingArgs)
+    val webClient: ClioWebClient = new ClioWebClient(
+      ClioClientConfig.ClioServer.clioServerHostName,
+      ClioClientConfig.ClioServer.clioServerPort,
+      ClioClientConfig.ClioServer.clioServerUseHttps,
+      ClioClientConfig.responseTimeout
+    )
+    val client: ClioClient = new ClioClient(webClient, IoUtil)
+
+    client
+      .instanceMain(args)
+      .fold({
+        case UsageOrHelpAsked(message) => {
+          println(message)
+          Future.successful(())
+        }
+        case ParsingError(error) => {
+          Future.failed(new RuntimeException(error))
+        }
+      }, identity)
+      .onComplete {
+        case Success(_) => sys.exit(0)
+        case Failure(ex) => {
```

Comment:
I like braces for anything > 1 line to avoid Python-esque "but where does this block end" adventures. Not really a problem here but I kept them for consistency.